File: src/components/fluid-simulation/fluid.ts
import { Grid } from "./grid";

export const Field = {
  U: "U",
  V: "V",
  S: "S", // smoke
} as const;

export type Field = (typeof Field)[keyof typeof Field];

export class Fluid {
  public u: Grid; // x-direction
  public v: Grid; // y-direction
  public s: Grid; // smoke
  private b: Grid; // blocks and obstacles
  // private p: Grid; // pressure
  private canvas: HTMLCanvasElement;

  private nextU: Grid; // x-direction
  private nextV: Grid; // y-direction
  private nextS: Grid; // smoke

  public squareSize: number;
  public gridWidth: number;
  public gridHeight: number;

  public minSquares: number;
  public nIterations: number;
  public deltaT: number;
  public overrelaxationCoefficient: number;

  public blockOffset: number;

  public get obstacles(): Grid {
    return this.b;
  }

  public constructor(
    canvas: HTMLCanvasElement,
    minSquares: number = 60,
    nIterations: number = 40,
    deltaT: number = 1 / 60,
    overrelaxationCoefficient: number = 1.7,
  ) {
    this.canvas = canvas;
    this.minSquares = minSquares;
    this.nIterations = nIterations;
    this.deltaT = deltaT;
    this.overrelaxationCoefficient = overrelaxationCoefficient;

    const h = Math.min(this.canvas.height, this.canvas.width);
    this.squareSize = h / this.minSquares;
    this.gridWidth = Math.ceil(this.canvas.width / this.squareSize);
    this.gridHeight = Math.ceil(this.canvas.height / this.squareSize);

    const n = 1;
    this.blockOffset = this.squareSize * n;

    this.u = new Grid(this.gridWidth + 1 + 2 * n, this.gridHeight + 2 * n);
    this.v = new Grid(this.gridWidth + 2 * n, this.gridHeight + 1 + 2 * n);
    this.b = new Grid(this.gridWidth + 2 * n, this.gridHeight + 2 * n);
    this.s = new Grid(this.gridWidth + 2 * n, this.gridHeight + 2 * n);

    this.nextU = new Grid(this.u.width, this.u.height);
    this.nextV = new Grid(this.v.width, this.v.height);
    this.nextS = new Grid(this.s.width, this.s.height);

    // initial smoke density
    this.randomizeSmoke();

    // the edges of the grid are obstacles
    this.b.fill(1);
    for (let i = 0; i < this.b.width; i++) {
      this.b.set(i, 0, 0);
      this.b.set(i, this.b.height - 1, 0);
    }

    for (let k = 0; k < this.b.height; k++) {
      this.b.set(0, k, 0);
      this.b.set(this.b.width - 1, k, 0);
    }
  }

  public simulate() {
    this.projection();
    this.advection();
  }

  public projection() {
    for (let l = 0; l < this.nIterations; l++) {
      this.solveDivergenceAll();
    }
  }

  public solveDivergenceAll() {
    this.b.forEach((val, i, k) => {
      if (val === 0) {
        this.v.set(i, k, 0);
        this.u.set(i, k, 0);
        return;
      }

      if (this.b.get(i - 1, k) === 0) {
        this.u.set(i, k, 0);
      }
      if (this.b.get(i, k - 1) === 0) {
        this.v.set(i, k, 0);
      }

      this.solveDivergence(i, k);
    });
  }

  public solveDivergence(i: number, k: number) {
    const b0 = this.b.get(i - 1, k);
    const b1 = this.b.get(i + 1, k);
    const b2 = this.b.get(i, k - 1);
    const b3 = this.b.get(i, k + 1);

    const b = b0 + b1 + b2 + b3;

    if (b === 0) {
      return;
    }

    const divergence =
      (this.getDivergence(i, k) * this.overrelaxationCoefficient) / b;

    this.u.set(i, k, (v) => v + divergence * b0);
    this.u.set(i + 1, k, (v) => v - divergence * b1);
    this.v.set(i, k, (v) => v + divergence * b2);
    this.v.set(i, k + 1, (v) => v - divergence * b3);
  }

  public getDivergence(i: number, k: number): number {
    return (
      this.u.get(i + 1, k) -
      this.u.get(i, k) +
      this.v.get(i, k + 1) -
      this.v.get(i, k)
    );
  }

  public advection() {
    // this.nextV = this.v;
    // this.nextU = this.u;
    // this.nextS = this.s;

    this.b.forEach((value, i, k) => {
      if (value === 0) {
        this.nextU.set(i, k, this.u.get(i, k));
        this.nextV.set(i, k, this.v.get(i, k));
        this.nextS.set(i, k, this.s.get(i, k));
        return;
      }

      // if (i === 5 && k === 5) {
      //   console.log(this.v.get(i, k));
      // }

      this.advectU(i, k);
      this.advectV(i, k);
      this.advectS(i, k);
    });

    this.b.forEach((_, i, k) => {
      this.u.set(i, k, this.nextU.get(i, k));
      this.v.set(i, k, this.nextV.get(i, k));
      this.s.set(i, k, this.nextS.get(i, k));
    });
  }

  public getGridPointFromCanvasPoint(
    point: [number, number],
    field?: Field,
  ): [number, number] {
    const x = Math.floor(
      (point[0] -
        ((field && field === Field.V) || field === Field.S
          ? this.squareSize / 2
          : 0) +
        this.blockOffset) /
        this.squareSize,
    );
    const y = Math.floor(
      (point[1] -
        ((field && field === Field.U) || field === Field.S
          ? this.squareSize / 2
          : 0) +
        this.blockOffset) /
        this.squareSize,
    );
    return [x, y] as const;
  }

  public getCanvasPointFromGridPoint(
    point: [number, number],
    field?: Field,
  ): [number, number] {
    const x =
      point[0] * this.squareSize +
      ((field && field === Field.V) || field === Field.S
        ? this.squareSize / 2
        : 0) -
      this.blockOffset;
    const y =
      point[1] * this.squareSize +
      ((field && field === Field.U) || field === Field.S
        ? this.squareSize / 2
        : 0) -
      this.blockOffset;
    return [x, y] as const;
  }

  public advectU(i: number, k: number) {
    if (this.b.get(i - 1, k) === 0) {
      this.nextU.set(i, k, this.u.get(i, k));
      return;
    }

    let [x, y] = this.getCanvasPointFromGridPoint([i, k], Field.U);

    const u = this.u.get(i, k);
    const vAvg = this.interpolate(x, y, Field.V);

    const previousX = x - u * this.deltaT;
    const previousY = y - vAvg * this.deltaT;

    this.nextU.set(i, k, this.interpolate(previousX, previousY, Field.U));
  }

  public advectV(i: number, k: number) {
    if (this.b.get(i, k - 1) === 0) {
      this.nextV.set(i, k, this.v.get(i, k));
      return;
    }

    let [x, y] = this.getCanvasPointFromGridPoint([i, k], Field.V);

    const v = this.v.get(i, k);
    const uAvg = this.interpolate(x, y, Field.U);

    const previousX = x - uAvg * this.deltaT;
    const previousY = y - v * this.deltaT;

    this.nextV.set(i, k, this.interpolate(previousX, previousY, Field.V));
  }

  private advectS(i: number, k: number) {
    let [x, y] = this.getCanvasPointFromGridPoint([i, k], Field.S);
    const u = this.interpolate(x, y, Field.U);
    const v = this.interpolate(x, y, Field.V);

    const previousX = x - u * this.deltaT;
    const previousY = y - v * this.deltaT;

    this.nextS.set(
      i,
      k,
      Math.max(this.interpolate(previousX, previousY, Field.S) - 2, 0),
    );
  }

  public interpolate(x: number, y: number, field: Field) {
    let fieldArr: Grid;

    switch (field) {
      case Field.U:
        fieldArr = this.u;
        break;
      case Field.V:
        fieldArr = this.v;
        break;
      case Field.S:
        fieldArr = this.s;
        break;
    }

    const [i, k] = this.getGridPointFromCanvasPoint([x, y], field);

    const [gridX, gridY] = this.getCanvasPointFromGridPoint([i, k], field);

    const xx = x - gridX;
    const yy = y - gridY;

    const wX = 1 - xx / this.squareSize;
    const wY = 1 - yy / this.squareSize;

    const newValBot =
      wX * fieldArr.get(i, k) + (1 - wX) * fieldArr.get(i + 1, k);

    const newValTop =
      wX * fieldArr.get(i, k + 1) + (1 - wX) * fieldArr.get(i + 1, k + 1);

    const newVal = wY * newValBot + (1 - wY) * newValTop;

    return newVal;
  }

  public clear() {
    this.u.fill(0);
    this.v.fill(0);
    this.s.fill(0);
  }

  public randomizeSmoke() {
    this.s.forEach((_, i, k) => {
      if (this.b.get(i, k) === 0) {
        return;
      }
      this.s.set(i, k, Math.random());
    });
  }
}

File: wasm/src/fluid.rs
use crate::grid::Grid;

#[derive(Debug, Clone)]
pub enum Field {
    U,
    V,
    S,
}
pub trait FluidSimulation {
    fn projection(&mut self);
    fn advection(&mut self);
    fn interpolate(&self, x: f64, y: f64, field: Field) -> f64;
    fn get_grid_indices_from_xy(&self, x: f64, y: f64, field: Option<&Field>) -> (i32, i32);
    fn get_xy_from_grid_indices(&self, x: i32, y: i32, field: Option<&Field>) -> (f64, f64);
}

#[derive(Debug, Clone)]
pub struct Fluid {
    pub u: Grid<f64>,      // velocity in x direction
    pub v: Grid<f64>,      // velocity in y direction
    pub b: Grid<u8>,       // obstacles
    pub s: Grid<f64>,      // smoke (density)
    pub next_u: Grid<f64>, // velocity in x direction
    pub next_v: Grid<f64>, // velocity in y direction
    pub next_s: Grid<f64>, // smoke (density)

    pub n: usize,
    pub square_size: f64,
    pub n_iterations: usize,
    pub delta_t: f64,
    pub overrelaxation_coefficient: f64,
    pub grid_width: usize,
    pub grid_height: usize,
    pub max_squares: usize,

    pub block_offset: f64,
}

impl Fluid {
    pub fn new(
        canvas: &web_sys::HtmlCanvasElement,
        min_squares: Option<usize>,
        n_iterations: Option<usize>,
        delta_t: Option<f64>,
        overrelaxation_coefficient: Option<f64>,
    ) -> Fluid {
        let max_squares = min_squares.unwrap_or(200);

        let h = u32::max(canvas.width(), canvas.height());

        let square_size: f64 = h as f64 / max_squares as f64;

        let grid_width = (canvas.width() as f64 / square_size).ceil() as usize;
        let grid_height = (canvas.height() as f64 / square_size).ceil() as usize;

        let n_iterations = n_iterations.unwrap_or(40);
        let delta_t = delta_t.unwrap_or(1_f64 / 60_f64);
        let overrelaxation_coefficient = overrelaxation_coefficient.unwrap_or(1.7);

        let n = 1;
        let block_offset = square_size * n as f64;

        let u = Grid::new(grid_width + 1 + 2 * n, grid_height + 2 * n);
        let v = Grid::new(grid_width + 2 * n, grid_height + 1 + 2 * n);
        let s = Grid::new(grid_width + 2 * n, grid_height + 2 * n);
        let b = Grid::new(grid_width + 2 * n, grid_height + 2 * n);

        let next_u = Grid::new(u.width(), u.height());
        let next_v = Grid::new(v.width(), v.height());
        let next_s = Grid::new(s.width(), s.height());

        Fluid {
            u,
            v,
            b,
            s,
            next_u,
            next_v,
            next_s,
            block_offset,
            grid_width,
            grid_height,
            max_squares,
            n,

            square_size,
            n_iterations,
            delta_t,
            overrelaxation_coefficient,
        }
    }

    pub fn fill_edges_with_obstacles(&mut self) {
        self.b.fill(1);

        for i in 0..self.b.width() {
            let i = i as i32;
            self.b.set(i, 0, 0);
            self.b.set(i, self.b.height() as i32 - 1, 0);
        }

        for k in 0..self.b.height() {
            let k = k as i32;
            self.b.set(0, k, 0);
            self.b.set(self.b.width() as i32 - 1, k, 0);
        }
    }

    pub fn solve_divergence_for_all(&mut self) {
        for i in 0..self.b.width() {
            for k in 0..self.b.height() {
                let i = i as i32;
                let k = k as i32;

                if self.b.get(i, k) == 0 {
                    self.u.set(i, k, 0.0);
                    self.v.set(i, k, 0.0);
                    continue;
                }
                if self.b.get(i - 1, k) == 0 {
                    self.u.set(i, k, 0.0);
                }
                if self.b.get(i, k - 1) == 0 {
                    self.v.set(i, k, 0.0);
                }

                self.solve_divergence(i, k);
            }
        }
    }

    fn solve_divergence(&mut self, i: i32, k: i32) {
        let b0 = self.b.get(i - 1, k);
        let b1 = self.b.get(i + 1, k);
        let b2 = self.b.get(i, k - 1);
        let b3 = self.b.get(i, k + 1);

        let b = b0 + b1 + b2 + b3;

        if b == 0 {
            return;
        }

        let divergence = (self.get_divergence(i, k) * self.overrelaxation_coefficient) / b as f64;

        self.u.update(i, k, |v| v + divergence * b0 as f64);
        self.u.update(i + 1, k, |v| v - divergence * b1 as f64);
        self.v.update(i, k, |v| v + divergence * b2 as f64);
        self.v.update(i, k + 1, |v| v - divergence * b3 as f64);
    }

    fn get_divergence(&self, i: i32, k: i32) -> f64 {
        self.u.get(i + 1, k) - self.u.get(i, k) + self.v.get(i, k + 1) - self.v.get(i, k)
    }

    fn advect_u(&mut self, i: i32, k: i32) {
        if self.b.get(i - 1, k) == 0 {
            self.next_u.set(i, k, self.u.get(i, k));
            return;
        }

        let (x, y) = self.get_xy_from_grid_indices(i, k, Some(&Field::U));

        let u = self.u.get(i, k);
        let v = self.interpolate(x, y, Field::V);

        let previous_x = x - u * self.delta_t;
        let previous_y = y - v * self.delta_t;

        let next_val = self.interpolate(previous_x, previous_y, Field::U);

        self.next_u.set(i, k, next_val);
    }

    fn advect_v(&mut self, i: i32, k: i32) {
        if self.b.get(i, k - 1) == 0 {
            self.next_v.set(i, k, self.v.get(i, k));
            return;
        }

        let (x, y) = self.get_xy_from_grid_indices(i, k, Some(&Field::V));

        let u = self.interpolate(x, y, Field::U);
        let v = self.v.get(i, k);

        let previous_x = x - u * self.delta_t;
        let previous_y = y - v * self.delta_t;

        let next_val = self.interpolate(previous_x, previous_y, Field::V);

        self.next_v.set(i, k, next_val);
    }

    fn advect_s(&mut self, i: i32, k: i32) {
        let (x, y) = self.get_xy_from_grid_indices(i, k, Some(&Field::S));

        let u = self.interpolate(x, y, Field::U);
        let v = self.interpolate(x, y, Field::V);

        let previous_x = x - u * self.delta_t;
        let previous_y = y - v * self.delta_t;

        let next_val = f64::max(
            self.interpolate(previous_x, previous_y, Field::S) / 1.1 - 0.1,
            0.0,
        );

        self.next_s.set(i, k, next_val);
    }

    pub fn clear(&mut self) {
        self.u.fill(0.0);
        self.v.fill(0.0);
        self.s.fill(0.0);
    }

    pub fn resize(&mut self, width: f64, height: f64) {
        let h = f64::max(width, height);
        let square_size: f64 = h / self.max_squares as f64;
        self.square_size = square_size;

        let grid_width = (width / square_size).ceil() as usize;
        let grid_height = (height / square_size).ceil() as usize;

        self.grid_width = grid_width;
        self.grid_height = grid_height;

        self.u
            .resize(grid_width + 2 * self.n + 1, grid_height + 2 * self.n);
        self.v
            .resize(grid_width + 2 * self.n, grid_height + 2 * self.n + 1);
        self.s
            .resize(grid_width + 2 * self.n, grid_height + 2 * self.n);
        self.b
            .resize(grid_width + 2 * self.n, grid_height + 2 * self.n);

        self.next_u.resize(self.u.width(), self.u.height());
        self.next_v.resize(self.v.width(), self.v.height());
        self.next_s.resize(self.s.width(), self.s.height());

        self.fill_edges_with_obstacles();
    }
}

impl FluidSimulation for Fluid {
    fn projection(&mut self) {
        for _ in 0..self.n_iterations {
            self.solve_divergence_for_all();
        }
    }

    fn advection(&mut self) {
        for i in 0..self.b.width() {
            for k in 0..self.b.height() {
                let i = i as i32;
                let k = k as i32;

                if self.b.get(i, k) == 0 {
                    self.next_u.set(i, k, self.u.get(i, k));
                    self.next_v.set(i, k, self.v.get(i, k));
                    self.next_s.set(i, k, self.s.get(i, k));
                    continue;
                }

                self.advect_u(i, k);
                self.advect_v(i, k);
                self.advect_s(i, k);
            }
        }

        self.u.swap(&mut self.next_u);
        self.v.swap(&mut self.next_v);
        self.s.swap(&mut self.next_s);
    }

    fn interpolate(&self, x: f64, y: f64, field: Field) -> f64 {
        let field_arr = match field {
            Field::U => &self.u,
            Field::V => &self.v,
            Field::S => &self.s,
        };

        let (i, k) = self.get_grid_indices_from_xy(x, y, Some(&field));
        let (grid_x, grid_y) = self.get_xy_from_grid_indices(i, k, Some(&field));

        let xx = x - grid_x;
        let yy = y - grid_y;

        let w_x = 1.0 - xx / self.square_size;
        let w_y = 1.0 - yy / self.square_size;

        let new_value_bot = w_x * field_arr.get(i, k) + (1.0 - w_x) * field_arr.get(i + 1, k);
        let new_value_top =
            w_x * field_arr.get(i, k + 1) + (1.0 - w_x) * field_arr.get(i + 1, k + 1);

        w_y * new_value_bot + (1.0 - w_y) * new_value_top
    }

    fn get_grid_indices_from_xy(&self, x: f64, y: f64, field: Option<&Field>) -> (i32, i32) {
        let i = (x - match field {
            Some(Field::V | Field::S) => self.square_size / 2.0,
            _ => 0.0,
        } + self.block_offset)
            / self.square_size;

        let k = (y - match field {
            Some(Field::U | Field::S) => self.square_size / 2.0,
            _ => 0.0,
        } + self.block_offset)
            / self.square_size;

        (i as i32, k as i32)
    }

    fn get_xy_from_grid_indices(&self, i: i32, k: i32, field: Option<&Field>) -> (f64, f64) {
        let x = i as f64 * self.square_size
            + match field {
                Some(Field::V | Field::S) => self.square_size / 2.0,
                _ => 0.0,
            }
            - self.block_offset;

        let y = k as f64 * self.square_size
            + match field {
                Some(Field::U | Field::S) => self.square_size / 2.0,
                _ => 0.0,
            }
            - self.block_offset;

        (x, y)
    }
}
